From 51cd243be5a24a322924a011ed875c32870e39d0 Mon Sep 17 00:00:00 2001
From: Om Narasimhan <omn@broadcom.com>
Date: Wed, 31 Oct 2012 11:20:02 -0700
Subject: [PATCH 616/761] PCIE : handling a hotplug controller bug

PLX 8624 does not handle command complete interrupts well. Enabling this
feature with Hotplug interrupts causes an interrupt storm, and effectively
leaves hot plug controller dysfunctional. This patch disables CCIE bit
for XLP

You need to pass pcie_ports=native in the kernel command line for XLP
hotplug to be functional.

Based on Broadcom SDK 2.3.

Signed-off-by: Om Narasimhan <omn@broadcom.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/pci/hotplug/pciehp_hpc.c |   12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/pci/hotplug/pciehp_hpc.c b/drivers/pci/hotplug/pciehp_hpc.c
index 9dd2c01..bb79598 100644
--- a/drivers/pci/hotplug/pciehp_hpc.c
+++ b/drivers/pci/hotplug/pciehp_hpc.c
@@ -830,7 +830,12 @@ int pcie_enable_notification(struct controller *ctrl)
 
 	mask = (PCI_EXP_SLTCTL_PDCE | PCI_EXP_SLTCTL_ABPE |
 		PCI_EXP_SLTCTL_MRLSCE | PCI_EXP_SLTCTL_PFDE |
-		PCI_EXP_SLTCTL_HPIE | PCI_EXP_SLTCTL_CCIE);
+		PCI_EXP_SLTCTL_HPIE );
+#if !defined CONFIG_NLM_XLP
+	/* Switching on CC int. enabling causes interrupt storm on XLP
+	 * Conditionally switching it off seems to be the only way of of it */
+	mask |= PCI_EXP_SLTCTL_CCIE;
+#endif
 
 	if (pcie_write_cmd(ctrl, cmd, mask)) {
 		ctrl_err(ctrl, "Cannot enable software notification\n");
@@ -844,8 +849,11 @@ static void pcie_disable_notification(struct controller *ctrl)
 	u16 mask;
 	mask = (PCI_EXP_SLTCTL_PDCE | PCI_EXP_SLTCTL_ABPE |
 		PCI_EXP_SLTCTL_MRLSCE | PCI_EXP_SLTCTL_PFDE |
-		PCI_EXP_SLTCTL_HPIE | PCI_EXP_SLTCTL_CCIE |
+		PCI_EXP_SLTCTL_HPIE |
 		PCI_EXP_SLTCTL_DLLSCE);
+#if !defined CONFIG_NLM_XLP
+	mask |= PCI_EXP_SLTCTL_CCIE;
+#endif
 	if (pcie_write_cmd(ctrl, 0, mask))
 		ctrl_warn(ctrl, "Cannot disable software notification\n");
 }
-- 
1.7.10.4

